### DISCUSSION TITLE
LDM Notes for March 9th, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5908] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-09.md

## Agenda

1. Ambiguity of `..` in collection expressions
2. `main` attributes
3. `nameof(param)`

**Comment by MgSam** [comment]
Not sure how `nameof(param)` went from an issue with little interest and even less support to being shipped for C# 11? What are the motivating cases for this feature? "Analyzers" seems like a poor answer- these are things that .01% of the .NET community writes, and even then, will likely only write once and not require a lot of maintenance. `nameof(Method.parameter)` is an unprecedented syntax and seems like it warrants more discussion.

I am more and more baffled over time by how you guys are prioritizing. It seems entirely based on back-channeling happening outside of public view now. This is a recipe for disaster over time, as you guys will ship more and more features that fulfill only niche situations yet make the language more complicated for everyone. As I've said repeatedly, the LDT hasn't even triaged many of the highest upvoted features in the Discussions, many of them have been there for years. 

I've followed the language design process since it went open source, and C# 11 is the first version of the language where I can say I don't care about almost any of the features in your working set outside of raw string literals.

**Reply by CyrusNajmabadi** [reply]
> What are the motivating cases for this feature? 

The nrt attributes that have to refer to parameter names. 

> nameof(Method.parameter) is an unprecedented syntax and seems like it warrants more discussion.

We do not support this syntax.

> As I've said repeatedly, the LDT hasn't even triaged many of the highest upvoted features in the Discussions, many of them have been there for years.

Which features are you referring to?

> I've followed the language design process since it went open source, and C# 11 is the first version of the language where I can say I don't care about almost any of the features in your working set outside of raw string literals.

Many of the features we're doing have some of hte highest vote counts we've ever seen on csharplang (and have positive reaction from many other sources as well). I get that things like 'field in auto-prop' may not be something you care about. But many others do, and we look at the entire ecosystem, and not hte individual when making decisions.

**Reply by Donaut** [reply]
Extension classes is one that is comes up from time to time:) (In different names) and discriminated unions too, very helpful features and not just a "if(param is null) { throw new Exception(); }" type of syntax sugar. I didn't read any discussion about duck typing but it's used by the compiler (await, foreach). But these things cannot be implemented without runtime changes so they starts from -1000 point. But the c# compiler has to be backward compatible so these things will probably never gonna get implemented in fact i bet that the c# compiler reached it's final form, and only new features gonna be accepted that can implemented by code generation.

**Reply by HaloFour** [reply]
@Donaut 

Structural types and extensions are a topic that the team is very interested in.  As are discriminated unions.  Such design is much more complicated and take a lot longer.  Those than require runtime changes take longer still.

Speaking of runtime changes, [abstract static members on interfaces](https://github.com/dotnet/csharplang/issues/4436) will be in preview in C# 11 and .NET 7, considered a first step towards structural typing that uses interfaces as concepts, so the teams are certainly willing to make runtime changes where it makes sense.

**Reply by Donaut** [reply]
@HaloFour 

Im pretty sceptical about it:). Lets see if is really gonna got implemented.

**Reply by 333fred** [reply]
> Im pretty sceptical about it:). Lets see if is really gonna got implemented.

It was implemented. You can go try it out right now in preview with NET 6.

**Reply by TonyValenti** [reply]
I am super excited about nameof param. I really like that for use with calleragrumentexprrssion.
